### PR TITLE
Kokkos + KokkosKernels Promotion To 4.3.0 (Test)

### PR DESCRIPTION
**Pull request by ndellingwood**
@trilinos/kokkos @trilinos/kokkos-kernels 

## Motivation
Test PR

### CODE REVIEW

**Review by lucbv — APPROVED** [review]

**Comment by ndellingwood** [comment]
Summary of cdash failures:

### tldr:
1. gcc/8.3.0 internal compiler errors with intrepid2 unit tests
2. Multiple jobs fail in the following MueLu tests:
  - MueLu_CreateOperatorTpetra_MPI_1, MueLu_CreateOperatorTpetra_MPI_4, MueLu_ParameterListInterpreterTpetra_MPI_1, MueLu_ParameterListInterpreterTpetra_MPI_4


MueLu_CreateOperatorTpetra_MPI_1 sample failure output:
```
default/Output/operator_solve_1_np1_tpetra: failed
==========================================================================================================================
===================================== Solve 5: LevelWrap, Fast Way, P, R =====================================
--- default/Output/operator_solve_5_np1_tpetra.gold_filtered	2024-03-26 16:39:07.065385262 -0600
+++ default/Output/operator_solve_5_np1_tpetra.out_filtered	2024-03-26 16:39:07.090385149 -0600
@@ -2,6 +2,32 @@
 
 Level 0
 Setup Smoother (MueLu::Ifpack2Smoother{type = RELAXATION})
+axpby(), unif information: AV = d, AtInputScalarTypeA = d, isConst = 0, isComplex = 0, AtInputScalarTypeA_nonConst = d, InternalTypeA = d
+, InternalTypeA_managed = v
+
+XMV = N6Kokkos4ViewIPPKdJNS_10LayoutLeftENS_6DeviceINS_6SerialENS_9HostSpaceEEENS_12MemoryTraitsILj0EEEEEE
+XMV::value_type = d
+XMV::const_data_type = PPKd
+XMV::non_const_data_type = PPd
+AtInputScalarTypeX = d
+isConst = 1
+isComplex = 0
+AtInputScalarTypeX_nonConst = d
+InternalTypeX = N6Kokkos4ViewIPPKdJNS_10LayoutLeftENS_6DeviceINS_6SerialENS_9HostSpaceEEENS_12MemoryTraitsILj1EEEEEE
+
+BV = d, AtInputScalarTypeB = d, isConst = 0, isComplex = 0, AtInputScalarTypeB_nonConst = d, InternalTypeB = d
+, InternalTypeB_managed = v
+
+YMV = N6Kokkos4ViewIPPdJNS_10LayoutLeftENS_6DeviceINS_6SerialENS_9HostSpaceEEENS_12MemoryTraitsILj0EEEEEE
+YMV::value_type = d
+YMV::const_data_type = PPKd
+YMV::non_const_data_type = PPd
+AtInputScalarTypeY = d
+isConst = 0
+isComplex = 0
+AtInputScalarTypeY_nonConst = d
+InternalTypeY = N6Kokkos4ViewIPPdJNS_10LayoutLeftENS_6DeviceINS_6SerialENS_9HostSpaceEEENS_12MemoryTraitsILj1EEEEEE
+
 smoother -> 
  relaxation: type = Symmetric Gauss-Seidel
  relaxation: sweeps = 1
```

### Details:

**Trilinos_PR_gcc-8.3.0**
- Multiple internal compilation errors with intrepid2 unit tests
```
g++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
```
- @trilinos/framework we may need help resolving (possibly lowering build level parallelism?)


**Trilinos_PR_gcc-8.3.0-debug**
- MueLu test failures: MueLu_CreateOperatorTpetra_MPI_1, MueLu_CreateOperatorTpetra_MPI_4, MueLu_ParameterListInterpreterTpetra_MPI_1, MueLu_ParameterListInterpreterTpetra_MPI_4
- Link to failures: https://trilinos-cdash.sandia.gov/viewTest.php?onlyfailed&buildid=1439074
- Next steps:
  - I was unable to reproduce the test failures in a local build using source override of the release candidates when running the tests individually, using the sems-gcc/8.3.0 stack and the generated cmake fragment posted on cdash; rebuilding snapshots
  - Look at getting access to retest manually on the failing node
  - @trilinos/muelu I may need some help with triage when I'm able to reproduce


**Trilinos_PR_gcc-8.3.0-serial**
- MueLu test failures: MueLu_CreateOperatorTpetra, MueLu_ParameterListInterpreterTpetra
- Link to failures: https://trilinos-cdash.sandia.gov/viewTest.php?onlyfailed&buildid=1439075


**Trilinos_PR_clang-11.0.1**
- MueLu test failures same as Trilinos_PR_gcc-8.3.0-debug: MueLu_CreateOperatorTpetra_MPI_1, MueLu_CreateOperatorTpetra_MPI_4, MueLu_ParameterListInterpreterTpetra_MPI_1, MueLu_ParameterListInterpreterTpetra_MPI_4
- ROL test timeout: ROL_example_PDE-OPT_navier-stokes_example_01_MPI_4
- Link to failures: https://trilinos-cdash.sandia.gov/viewTest.php?onlyfailed&buildid=1439077
- Next steps:
  - Same as with Trilinos_PR_gcc-8.3.0-debug
  - I'm curious to try retesting with the failing ROL test marked to RUN_SERIAL, the timeout might indicate some resource contention due to overlapping tests


**Trilinos_PR_intel-2021.3**
- MueLu test failures: MueLu_CreateOperatorTpetra_MPI_1, MueLu_CreateOperatorTpetra_MPI_4, MueLu_ParameterListInterpreterTpetra_MPI_1, MueLu_ParameterListInterpreterTpetra_MPI_4
- Link to failures: https://trilinos-cdash.sandia.gov/viewTest.php?onlyfailed&buildid=1439445

**Comment by ndellingwood** [comment]
Reproduced the MueLu test failures; may test fixes here or in a separate PR with the final snapshots

File: packages/kokkos-kernels/blas/src/KokkosBlas1_axpby.hpp [hunk 1]
```diff
@@ -17,164 +17,312 @@
...
+  using InternalTypeX = typename AxpbyTraits::InternalTypeX;
+  using InternalTypeB = typename AxpbyTraits::InternalTypeB;
+  using InternalTypeY = typename AxpbyTraits::InternalTypeY;
+
+  // **********************************************************************
+  // Perform compile time checks and run time checks.
+  // **********************************************************************
+  AxpbyTraits::performChecks(a, X, b, Y);
+#ifdef HAVE_KOKKOSKERNELS_DEBUG
+  AxpbyTraits::printInformation(std::cout, "axpby(), unif information");
```
Comment (ndellingwood) [inline packages/kokkos-kernels/blas/src/KokkosBlas1_axpby.hpp:77]: This line is triggering the MueLu test failures in the host builds; commenting it out allowed tests to pass MueLu_ParameterListInterpreterTpetra and MueLu_CreateOperatorTpetra to pass in the gcc/8.3.0 serial debug build

**Comment by ndellingwood** [comment]
MueLu failures resolved for me with https://github.com/kokkos/kokkos-kernels/pull/2157

Will likely close this PR without additional testing and replace with a new PR containing the finalized 4.3 snapshots

In addition, I'd like to preemptively add the RUN_SERIAL marking for `ROL_example_PDE-OPT_navier-stokes_example_01_MPI_4` to avoid timeouts
